File: RDE/data/SDToolbox/MATLAB/Demo/demo_oblique.m
% Shock and Detonation Toolbox Demo Program
% 
% Calculates shock polar using FROZEN post-shock state based the initial gas properties 
% and the shock speed. Plots shock polar using three different sets of coordinates.
%  
% ################################################################################
% Theory, numerical methods and applications are described in the following report:
% 
%     Numerical Solution Methods for Shock and Detonation Jump Conditions, S.
%     Browne, J. Ziegler, and J. E. Shepherd, GALCIT Report FM2006.006 - R3,
%     California Institute of Technology Revised September, 2018.
% 
% Please cite this report and the website if you use these routines. 
% 
% Please refer to LICENCE.txt or the above report for copyright and disclaimers.
% 
% http://shepherd.caltech.edu/EDL/PublicResources/sdt/
% 
% ################################################################################ 
% Updated September 2018
% Tested with: 
%     MATLAB 2017b and 2018a, Cantera 2.3 and 2.4
% Under these operating systems:
%     Windows 8.1, Windows 10, Linux (Debian 9)
%%
clear;clc;
disp('demo_oblique') 
%%
% set the initial state and compute properties
P1 = 100000; T1 = 300; 
q = 'H2:2 O2:1 N2:3.76';    
mech = 'Mevel2017.cti'; 
gas1 = Solution(mech);
set(gas1,'Temperature',T1,'Pressure',P1,'MoleFractions',q);
rho1 = density(gas1);
a1 = soundspeed_fr(gas1);
%%
% Set freestream velocity and find limiting speeds
U = 1500.;
beta_min = asin(a1/U);  %Mach angle
wmin = a1 + 1.;  %start just above sound speed
% check to see if shock speed is above sound speed
if (U < a1)
     exit;
end
% set maximum normal speed
wmax = U;
% initialize variables for plotting
rho2 = 0.;
w1 = 0.;
w2 = 0.;
u2 = 0.;
v = 0.;
a2 = 0.;
P2 = 0.;
beta = 0.;
theta = 0.;
n = 1;
%%
% compute shock jump conditions over range from minimum to maximum normal
% speeds.  Adjust the increment to get a smooth output curve. 
for w = wmin:5:wmax
    [gas] = PostShock_fr(w, P1, T1, q, mech);
%    [gas] = PostShock_eq(w, P1, T1, q, mech);  %will not converge at low
%    temperature.
    rho2(n) = density(gas);
    a2(n) = soundspeed_fr(gas);
    ratio = rho1/rho2(n);
    w1(n) = w;
    w2(n) = w*ratio;
    P2(n) = pressure(gas);
    beta(n) = asin(w/U);
    v(n) = U*cos(beta(n));
    theta(n) = beta(n) - atan(w2(n)/sqrt(U^2-w^2));
    u2(n) = sqrt(w2(n)^2 + v(n)^2);
    n = n + 1;
end
% pick up normal shock as last point
    w = U;
    [gas] = PostShock_fr(w, P1, T1, q, mech);
    rho2(n) = density(gas);
    a2(n) = soundspeed_fr(gas);
    ratio = rho1/rho2(n);
    w1(n) = w;
    w2(n) = w*ratio;
    P2(n) = pressure(gas);
    beta(n) = asin(w/U);
    v(n) = U*cos(beta(n));
    theta(n) = beta(n) - atan(w2(n)/sqrt(U^2-w^2));
    u2(n) = sqrt(w2(n)^2 + v(n)^2);
    %%  post process output 
    % compute Mach number 
    M2 = u2./a2;
    % find sonic point
    beta_star = interp1(M2,beta,1.0);
    theta_star = interp1(beta,theta,beta_star);
    % find maximum deflection estimate
    theta_max_guess = max(theta);
    beta_max_guess = interp1(theta,beta,theta_max_guess);
    fun = @(x)interp1(beta,theta,x,'pchip');
    beta_max = fminbnd(@(x)-fun(x), .9*beta_max_guess, 1.1*beta_max_guess);
    theta_max = interp1(beta, theta, beta_max, 'pchip');
    % compute velocity components downstream in lab frame
    u2x = u2.*cos(theta);
    u2y = u2.*sin(theta);
    %% create plots
% plot pressure-deflection polar
	figure(1); clf;
	plot(180*theta(:)/pi,P2(:)/1e6,'k:','LineWidth',2);  
%	axis([0 maxtheta 0 maxP]);
	title(['Shock Polar Air, free-stream speed ',num2str(U,5),' m/s'],'FontSize', 12);
	xlabel('deflection angle (deg)','FontSize', 12);
	ylabel('pressure (MPa)','FontSize', 12);
	set(gca,'FontSize',12,'LineWidth',2);
% plot pressure-deflection polar
	figure(2); clf;
	plot(180*theta(:)/pi,180*beta(:)/pi,'k:','LineWidth',2);  
	title(['Shock Polar Air, free-stream speed ',num2str(U,5),' m/s'],'FontSize', 12);
	xlabel('deflection angle (deg)','FontSize', 12);
	ylabel('wave angle (deg)','FontSize', 12);
	set(gca,'FontSize',12,'LineWidth',2);
% plot velocity polar
	figure(3); clf;
	plot(u2x(:),u2y(:),'k:','LineWidth',2);  
	title(['Shock Polar Air, free-stream speed ',num2str(U,5),' m/s'],'FontSize', 12);
	xlabel('u2x (m/s)','FontSize', 12);
	ylabel('u2y (m/s)','FontSize', 12);
	set(gca,'FontSize',12,'LineWidth',2);

File: RDE/data/SDToolbox/Python3/demo/demo_oblique.py
"""
Shock and Detonation Toolbox Demo Program

Calculates shock polar using FROZEN post-shock state based the initial gas properties 
and the shock speed. Plots shock polar using three different sets of coordinates.
Saves archive of numpy arrays for beta, theta, and M_2.

################################################################################
Theory, numerical methods and applications are described in the following report:

SDToolbox - Numerical Tools for Shock and Detonation Wave Modeling,
Explosion Dynamics Laboratory, Contributors: S. Browne, J. Ziegler,
N. Bitter, B. Schmidt, J. Lawson and J. E. Shepherd, GALCIT
Technical Report FM2018.001 Revised January 2021.

Please cite this report and the website if you use these routines. 

Please refer to LICENCE.txt or the above report for copyright and disclaimers.

http://shepherd.caltech.edu/EDL/PublicResources/sdt/

################################################################################ 
Updated September 2018
Tested with: 
    Python 3.5 and 3.6, Cantera 2.3 and 2.4
Under these operating systems:
    Windows 8.1, Windows 10, Linux (Debian 9)s
"""
import cantera as ct
from sdtoolbox.thermo import soundspeed_fr
from sdtoolbox.postshock import PostShock_fr #, PostShock_eq
import numpy as np
from sys import exit
from numpy import interp # directly takes a query point
from scipy.interpolate import pchip # creates a PCHIP interpolation function that can then be given a query point
from scipy.optimize import fminbound
import matplotlib.pyplot as plt

# Suppress the 'divide by zero' warning encountered for the last data point in the arctan calls
# (where U = w) so that we can keep the code as similar as possible to the MATLAB equivalent
# Numpy actually handles this case and returns the correct value for the limit arctan(inf)
import warnings
warnings.filterwarnings("ignore", category=RuntimeWarning)

##
# set the initial state and compute properties
P1 = 100000; T1 = 300 
q = 'H2:2 O2:1 N2:3.76'   
mech = 'Mevel2017.cti'
gas1 = ct.Solution(mech)
gas1.TPX = T1,P1,q
rho1 = gas1.density
a1 = soundspeed_fr(gas1)

##
# Set freestream velocity and find limiting speeds
U = 1500.
beta_min = np.arcsin(a1/U)  # Mach angle
wmin = a1 + 1.  # start just above sound speed
# check to see if shock speed is above sound speed
if (U < a1):
    exit('Shock speed below sound speed. Exiting.')


# set maximum normal speed
wmax = U

# initialize variables for plotting
rho2 = []; w1 = []; w2 = []; u2 = []; v = []; a2 = []; P2 = []; beta = []; theta = [];

##
# compute shock jump conditions over range from minimum to maximum normal
# speeds.  Adjust the increment to get a smooth output curve.
wstep = 5
nsteps = int((wmax-wmin)/wstep)
for w in np.linspace(wmin,wmax,num=nsteps):
    gas = PostShock_fr(w, P1, T1, q, mech)
    # gas = PostShock_eq(w, P1, T1, q, mech) #will not converge at low temperature.
    rho2.append(gas.density)
    a2.append(soundspeed_fr(gas))
    ratio = rho1/rho2[-1]
    w1.append(w)
    w2.append(w*ratio)
    P2.append(gas.P)
    beta.append(np.arcsin(w/U))
    v.append(U*np.cos(beta[-1]))
    theta.append(beta[-1] - np.arctan(w2[-1]/np.sqrt(U**2-w**2)))
    u2.append(np.sqrt(w2[-1]**2 + v[-1]**2))

## pick up normal shock as last point
w = U
gas = PostShock_fr(w, P1, T1, q, mech)
rho2[-1] = gas.density
a2[-1] = soundspeed_fr(gas)
ratio = rho1/rho2[-1]
w1[-1] = w
w2[-1] = w*ratio
P2[-1] = gas.P
beta[-1] = np.arcsin(w/U)
v[-1] = U*np.cos(beta[-1])
theta[-1] = beta[-1] - np.arctan(w2[-1]/np.sqrt(U**2-w**2))
u2[-1] = np.sqrt(w2[-1]**2 + v[-1]**2)

## post process output 
# compute Mach number 
u2 = np.asarray(u2); a2 = np.asarray(a2); P2 = np.asarray(P2)
beta = np.asarray(beta); theta = np.asarray(theta)
M2 = u2/a2

np.savez('BTM',beta=beta,theta=theta,M2=M2)

# find sonic point
# Note: numpy.interp takes a query point directly, like MATLAB's interp1. 
# However, it requires the x-coordinates of the data points to be monotonically increasing.
# Make sure x-data gets sorted to be monotonically increasing before attempting any interpolation
inds = M2.argsort()
M2 = M2[inds]; beta = beta[inds]; theta = theta[inds];
beta_star = interp(1.0,M2,beta)

inds = beta.argsort()
M2 = M2[inds]; beta = beta[inds]; theta = theta[inds];
theta_star = interp(beta_star,beta,theta)

# find maximum deflection estimate
theta_max_guess = max(theta)
inds = theta.argsort()
M2 = M2[inds]; beta = beta[inds]; theta = theta[inds];
beta_max_guess = interp(theta_max_guess,theta,beta)

inds = beta.argsort()
M2 = M2[inds]; beta = beta[inds]; theta = theta[inds];
fun = pchip(beta,theta)
beta_max = fminbound(lambda x: -fun(x), .9*beta_max_guess, 1.1*beta_max_guess)
theta_max = fun(beta_max)

## compute velocity components downstream in lab frame
u2x = u2*np.cos(theta)
u2y = u2*np.sin(theta)

## create plots

# plot pressure-deflection polar
plt.figure(1)
plt.plot(180*theta/np.pi,P2/1e6,'k:',linewidth=2)
plt.title('Shock Polar Air, free-stream speed '+str(round(U,5))+' m/s',fontsize=12)
plt.xlabel('deflection angle (deg)',fontsize= 12)
plt.ylabel('pressure (MPa)',fontsize= 12)
plt.tick_params(labelsize=12)
plt.xlim(xmin=0)
plt.ylim(ymin=0)


# plot pressure-deflection polar
plt.figure(2)
plt.plot(180*theta/np.pi,180*beta/np.pi,'k:',linewidth=2)  
plt.title('Shock Polar Air, free-stream speed '+str(round(U,5))+' m/s',fontsize=12)
plt.xlabel('deflection angle (deg)',fontsize=12)
plt.ylabel('wave angle (deg)',fontsize=12)
plt.tick_params(labelsize=12)
plt.xlim(xmin=0)
plt.ylim(0,90)

# plot velocity polar
plt.figure(3)
plt.plot(u2x,u2y,'k:',linewidth=2)  
plt.title('Shock Polar Air, free-stream speed '+str(round(U,5))+' m/s',fontsize=12)
plt.xlabel('u2x (m/s)',fontsize=12)
plt.ylabel('u2y (m/s)',fontsize=12)
plt.tick_params(labelsize=12)
plt.ylim(ymin=0)

plt.show()
